rancher/rancher
# Add opsgenie notifer support (#18164)

This pull request is solving #18164

This pull request is WIP since it is waiting for [https://github.com/rancher/types/pull/931](https://github.com/rancher/types/pull/931)

Since UI/Web development is beyond my possibilities this is about the API side only. Would be great to find someone to integrate this with the UI.

* Most of the config stuff was already existing.
* I had to fix `opsgenie_api_host` to `opsgenie_api_url` and `api_host` to `api_url`.
* I have streamlined opsgenie to be a single word. This results in renaming all `[oO]psGenie` into `[oO]psgenie`

Please let me know if you need additional informations. Also I would love to see this backported to 2.2 but don't know how your process for backports looks like.

Kind regards
Wulf

## Steps to verify

* Create a Notifier with the `opsgenieConfig` field with value:

```json
{
    "type":"/v3/schemas/opsgenieConfig",
    "api_key":"my-api-key",
    "region": "eu",
    "defaultRecipient": "my-team",
    "tags": "my-tag,my-tag2"
}
```

* Configure a Alert-Group to use this notifier
* Check the generated alertmanager config to be:

```yaml
global:
  resolve_timeout: 5m
  http_config: {}
  smtp_hello: localhost
  slack_api_url: <secret>
  pagerduty_url: https://events.pagerduty.com/v2/enqueue
  hipchat_api_url: https://api.hipchat.com/
  opsgenie_api_url: https://api.opsgenie.com/
  wechat_api_url: https://qyapi.weixin.qq.com/cgi-bin/
  victorops_api_url: https://alert.victorops.com/integrations/generic/20131114/alert/
route:
  receiver: rancherlabs
  routes:
  - receiver: p-stmzn:projectalert-workload-alert
    group_by:
    - group_id
    match:
      group_id: p-stmzn:projectalert-workload-alert
    continue: true
    routes:
    - group_by:
      - rule_id
      - workload_namespace
      - workload_name
      - workload_kind
      match:
        rule_id: p-stmzn:projectalert-workload-alert_less-than-half-workload-available
      continue: true
    - group_by:
      - rule_id
      match:
        rule_id: p-stmzn:projectalert-workload-alert_memory-close-to-resource-limited
      continue: true
    group_wait: 3m
    group_interval: 3m
    repeat_interval: 1h
  group_wait: 1m
  group_interval: 10s
  repeat_interval: 1h
receivers:
- name: rancherlabs
  slack_configs:
  - send_resolved: false
    http_config: {}
    api_url: <secret>
    channel: '#alert'
    username: '{{ template "slack.default.username" . }}'
    color: '{{ if eq .Status "firing" }}danger{{ else }}good{{ end }}'
    title: '{{ template "slack.default.title" . }}'
    pretext: '{{ template "slack.default.pretext" . }}'
    text: '{{ template "slack.default.text" . }}'
    footer: '{{ template "slack.default.footer" . }}'
    fallback: '{{ template "slack.default.fallback" . }}'
    callback_id: '{{ template "slack.default.callbackid" . }}'
    icon_emoji: '{{ template "slack.default.iconemoji" . }}'
    icon_url: '{{ template "slack.default.iconurl" . }}'
- name: p-stmzn:projectalert-workload-alert
  opsgenie_configs:
  - send_resolved: true
    http_config: {}
    api_key: <secret>
    api_url: https://api.eu.opsgenie.com/
    message: '{{ template "rancher.title" . }}'
    description: '{{ template "opsgenie.default.description" . }}'
    source: '{{ template "opsgenie.default.source" . }}'
    teams: my-team
    tags: my-tag,my-tag2
    priority: '{{ if eq (index .Alerts 0).Labels.severity "critical" }}P1{{ else if
      eq (index .Alerts 0).Labels.severity "warning" }}P2{{ else }}P3{{ end }}'
templates:
- /etc/alertmanager/config/notification.tmpl
```
